vmaas_sync/repo_based.go

```diff
@@ -0,0 +1,113 @@
+package vmaas_sync //nolint:golint,stylecheck
+import (
+	"app/base/database"
+	"app/base/models"
+	"app/base/utils"
+	"context"
+	"github.com/RedHatInsights/patchman-clients/vmaas"
+	"github.com/antihax/optional"
+	"time"
+)
+
+const LastEvalRepoBased = "last_eval_repo_based"
+
+func getCurrentRepoBasedInventoryIDs() (*[]string, error) {
+	lastRepoBaseEval, err := getLastRepobasedEvalTms()
+	if err != nil {
+		return nil, err
+	}
+
+	updateRepos, err := getUpdatedRepos(lastRepoBaseEval)
+	if err != nil {
+		return nil, err
+	}
+
+	inventoryIDs, err := getRepoBasedInventoryIDs(updateRepos)
+	if err != nil {
+		return nil, err
+	}
+
+	updateRepoBaseEvalTimestamp(time.Now())
+
+	return inventoryIDs, nil
+}
+
+func getLastRepobasedEvalTms() (*time.Time, error) {
+	var timestamps []*time.Time
+	err := database.Db.Model(&models.TimestampKV{}).
+		Where("name = ?", LastEvalRepoBased).
+		Pluck("value", &timestamps).Error
+	if err != nil {
+		return nil, err
+	}
+
+	if len(timestamps) == 0 {
+		return nil, nil
+	}
+
+	return timestamps[0], nil
+}
+
+func updateRepoBaseEvalTimestamp(value time.Time) {
+	ts := value.Format(time.RFC3339)
+	err := updateRepoBaseEvalTimestampStr(ts)
+	if err != nil {
+		utils.Log("err", err.Error()).Error("unable to update repo-based eval. timestamp")
+	}
+}
+
+func updateRepoBaseEvalTimestampStr(value string) error {
+	err := database.Db.Exec("INSERT INTO timestamp_kv (name, value) values (?, ?)"+
+		"ON CONFLICT (name) DO UPDATE SET value = ?", LastEvalRepoBased, value, value).Error
+	return err
+}
+
+func getRepoBasedInventoryIDs(repos *[]string) (*[]string, error) {
```

Comment:
Pointer to slice

Reply:
Yes, and also in other places. Is there any reason not to use it any more?

Reply:
In go everything is copy-by-value.
You'd be right to assume that passing a slice to function would copy it, except that this only applies to fixed size array.

Passing a slice into a function only passes the slice header by value, and the actual backing array is not copied.

By creating a pointer to slice you're creating an 8byte pointer to 24byte header, which contains a pointer to actual data.

If I'm not mistaken, creating this pointer means that escape analysis will conclude that the slice will have to be allocated on heap instead of stack.

So, we have
stack -> heap -> heap
[ptr] -> [ptr, len, cap] -> [... data]

instead of 
stack -> heap
[ptr, len, cap] -> [ ... data]

Also ergonomics, we don't have to type all the dereferences and parentheses.


From what I understand, the only necessary place to use pointer to slice is when you want to modify the slice length/capacity in a function, and have that reflected in caller.